Frontend/HsParser.hs

```diff
@@ -281,12 +281,33 @@ pTransPat (PsdAppPat p1 p2) = do
   p2' <- pTransPat p2
   return $ HsConPat dc (xs ++ [p2']) -- NOTE: this is O(n^2) because of the (++), could be O(n)
 
+-- | Wrap the contents of a monad into a singleton array
+one :: Monad f => f a -> f [a]
+one x = x >>= (\x' -> return [x'])
+
 -- | Parse a pattern
 pPat :: PsM PsPat
 pPat = pTransPat <$> pPatMain >>= \case
   Nothing -> empty
   Just p  -> return p
 
+-- | Parse guards
+pGuards :: PsM [PsGuard]
+pGuards = indent (symbol "|" *> sepBy1 pGuard (symbol ", "))
+
+-- | Parse a list of pattern guards
+pGuard :: PsM PsGuard
+pGuard =  HsGuard <$> pPat <* symbol "<-" <*> pTerm
+
+-- | Parse an arrow and right hand side term
+pRhs :: PsM PsTerm
+pRhs = symbol "->" *> pTerm
```

Comment:
Best if you inline this definition. I know it looks repetitive but it doesn't make much sense as a standalone function.